### ISSUE TITLE
Playback failing on device when moving from 4.7.15 to 4.8.0

### ISSUE THREAD

**Issue by MikeKav** [shaka-project/shaka-player#6844]
**Have you read the [Tutorials](http://shaka-player-demo.appspot.com/docs/api/tutorial-welcome.html)?**
Yes

**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**
Yes

**If the question is related to FairPlay, have you read the tutorial?**
N/A

**What version of Shaka Player are you using?**
Last working: 4.7.15
Unworking: 4.8.0 and 4.9.0

**What browser and OS are you using?**
EE/YouView TV Box Pro Set Top Box (https://www.bt.com/help/tv/learn-about-tv/bt-tv-boxes), Sagemcom hardware running WPE browser version 2.42.5 (slightly higher versions available soon)
YouViewHTML/1.0 AppleWebKit/605.1.15 (Sagemcom; RTIW387; RTIW387.002.X; CDS/0.10.116; API/4.0.0; PS/4.1.144) (+DVR+HTML+IPCMC+UHD+DASH+DRM+MSE)

**Please ask your question**
We've had a issue that playback on this STB has worked from 4.2.10 until 4.7.15, but we are experiencing failure to see/hear video/audio content with DASH/PlayReady material. Clear DASH has continued to work on both 4.8.0 and 4.9.0.

The same code is playing video and audio on Edge on all three of the above versions.

Console output seems to indicate the functionality is semi-working: There is a lot of output on setup, but then the player gets stuck in a loop repeating the below constantly:
[Debug] (video:8) – "timeNeeded=10" (shaka-player.compiled.debug.js, line 143)
[Debug] (video:8) – "update_:" – "presentationTime=0" – "bufferedAhead=10" (shaka-player.compiled.debug.js, line 143)
[Debug] (video:8) – "buffering goal met" (shaka-player.compiled.debug.js, line 143)
[Debug] (video:8) – "updating in 0.5 seconds" (shaka-player.compiled.debug.js, line 143)
[Debug] (audio:2) – "timeNeeded=10.005333333333333" (shaka-player.compiled.debug.js, line 143)
[Debug] (audio:2) – "update_:" – "presentationTime=0" – "bufferedAhead=10.005333" (shaka-player.compiled.debug.js, line 143)
[Debug] (audio:2) – "buffering goal met" (shaka-player.compiled.debug.js, line 143)
[Debug] (audio:2) – "updating in 0.5 seconds" (shaka-player.compiled.debug.js, line 143)

We do have AppleWebKit in our user agent string, but have not patched in an is[Device] API as used in [platform.js](https://github.com/shaka-project/shaka-player/blob/main/lib/util/platform.js#L294) as we have until now been able to use configuration switches to set behavior to a working approach.

Up until 4.7.14 the only changes we've needed have been 
player.configure('streaming.useNativeHlsOnSafari', false); << This was used originally in 4.2.10 to switch our devices DASH streaming from native engine to MSE/EME
and currently for audio track switching where codecs are different we now need to add:
player.configure('mediaSource.codecSwitchingStrategy', shaka.config.CodecSwitchingStrategy.RELOAD)

We'd appreciate any ideas for areas to look at to find the problem, we've done a comparison of the new fields in config between 4.7.15 and 4.8.0 and tried a few varying settings with no visible change to behaviours

Results from https://shaka-player-demo.appspot.com/support.html: 

YouViewHTML/1.0 AppleWebKit/605.1.15 (Sagemcom; RTIW387; RTIW387.002.X; CDS/0.10.116; API/4.0.0; PS/4.1.144) (+DVR+HTML+IPCMC+UHD+DASH+DRM+MSE)
v4.9.5

{
  "manifest": {
    "application/dash+xml": true,
    "video/vnd.mpeg.dash.mpd": true,
    "application/x-mpegurl": true,
    "application/vnd.apple.mpegurl": true,
    "application/vnd.ms-sstr+xml": true,
    "application/x-offline-manifest": true
  },
  "media": {
    "video/mp4; codecs=\"avc1.42E01E\"": true,
    "video/mp4": true,
    "video/mp4; codecs=\"avc3.42E01E\"": true,
    "video/mp4; codecs=\"hev1.1.6.L93.90\"": true,
    "video/mp4; codecs=\"hvc1.1.6.L93.90\"": true,
    "video/mp4; codecs=\"hev1.2.4.L153.B0\"; eotf=\"smpte2084\"": true,
    "video/mp4; codecs=\"hvc1.2.4.L153.B0\"; eotf=\"smpte2084\"": true,
    "video/mp4; codecs=\"vp9\"": false,
    "video/mp4; codecs=\"vp09.00.10.08\"": true,
    "video/mp4; codecs=\"av01.0.01M.08\"": false,
    "video/mp4; codecs=\"dvh1.20.01\"": false,
    "audio/mp4; codecs=\"mp4a.40.2\"": true,
    "audio/mp4": true,
    "audio/mp4; codecs=\"ac-3\"": true,
    "audio/mp4; codecs=\"ec-3\"": true,
    "audio/mp4; codecs=\"ac-4.02.01.01\"": false,
    "audio/mp4; codecs=\"opus\"": true,
    "audio/mp4; codecs=\"flac\"": false,
    "audio/mp4; codecs=\"dtsc\"": false,
    "audio/mp4; codecs=\"dtse\"": false,
    "audio/mp4; codecs=\"dtsx\"": false,
    "video/webm; codecs=\"vp8\"": false,
    "video/webm": true,
    "video/webm; codecs=\"vp9\"": false,
    "video/webm; codecs=\"vp09.00.10.08\"": true,
    "audio/webm; codecs=\"vorbis\"": false,
    "audio/webm": true,
    "audio/webm; codecs=\"opus\"": true,
    "video/mp2t; codecs=\"avc1.42E01E\"": true,
    "video/mp2t": true,
    "video/mp2t; codecs=\"avc3.42E01E\"": true,
    "video/mp2t; codecs=\"hvc1.1.6.L93.90\"": true,
    "video/mp2t; codecs=\"mp4a.40.2\"": true,
    "video/mp2t; codecs=\"ac-3\"": true,
    "video/mp2t; codecs=\"ec-3\"": true,
    "text/vtt": true,
    "application/mp4; codecs=\"wvtt\"": true,
    "application/mp4": true,
    "application/ttml+xml": true,
    "application/mp4; codecs=\"stpp\"": true,
    "audio/aac": true,
    "audio/ac3": true,
    "audio/ec3": true,
    "audio/mpeg": true
  },
  "drm": {
    "org.w3.clearkey": null,
    "com.widevine.alpha": null,
    "com.microsoft.playready": {
      "persistentState": false,
      "encryptionSchemes": [
        "cenc",
        "cbcs"
      ],
      "videoRobustnessLevels": [
        "150",
        "2000",
        "3000"
      ],
      "audioRobustnessLevels": [
        "150",
        "2000",
        "3000"
      ]
    },
    "com.microsoft.playready.recommendation": null,
    "com.chromecast.playready": null,
    "com.apple.fps.1_0": null,
    "com.apple.fps": null
  },
  "hardwareResolution": {
    "width": null,
    "height": null
  },
  "offline": true
}

**Comment by avelad** [comment]
We do not have that device, I recommend that you look at the recent commit history in DrmEngine (https://github.com/shaka-project/shaka-player/commits/main/lib/media/drm_engine.js) and if you find the commit that fails, we will be able to help you... Sorry!

**Comment by MikeKav** [comment]
Hi, I tried a few versions!

Using:
https://github.com/shaka-project/shaka-player/pull/5897 (feat: Add preload system to player)
git checkout -b test-branch-489b11a 489b11a
python3 build/all.py --force
PlayReady Content fails to play

https://github.com/shaka-project/shaka-player/pull/6189 (feat!: Remove com.adobe.primetime keysystem)
git checkout -b test-branch-47602c6 47602c6
python3 build/all.py --force
PlayReady Content plays

To check I didn't make an error I then:
git checkout test-branch-489b11a
python3 build/all.py --force

The version numbers reported seemed a bit broken,  but I assume from this test it's something in the changes in https://github.com/shaka-project/shaka-player/pull/5897 that is causing the issue for me?

**Comment by MikeKav** [comment]
I'll also add that I had not yet moved to using the player.attach() method, but now I have, with the same results

**Comment by avelad** [comment]
The PR is huge, but there is no change that affects the drm.

**Comment by avelad** [comment]
The change is very big, if you can debug more, and find where the error is, we are happy to fix it :)

**Comment by MikeKav** [comment]
@shaka-bot reopen

I've not had time to work on this until now, today I've tried the newer versions as released since I first posted.
4.17.15 (/)
4.8.0 (x), 4.8.20 (x)
4.9.0 (x), 4.9.19 (x)
4.10.0 (x), 4.10.7 (x)

**Comment by avelad** [comment]
@MikeKav We do not have that device, we can leave the issue open, but you will have to debug the error yourself, sorry!

**Comment by MikeKav** [comment]
Hi @avelad , I do understand (as a matter of interest, do you have a UK based lab?). I've re-opened to provide some further details, hopefully someone might see something I'm doing and be able to help advise.
I have checked some different content types, and can replicate the issue on public available content and keys, which is:
                "url": "https://media.axprod.net/TestVectors/v7-MultiDRM-SingleKey/Manifest_1080p.mpd",

                    "drm": {
                        "servers": {
                            "com.microsoft.playready":"https://test.playready.microsoft.com/service/rightsmanager.asmx?cfg=(kid:9eb4050d-e44b-4802-932e-27d75083e266,contentkey:FmY0xnWCPCNaSpRG+tUuTQ==,sl:150)"
                        }
                    },

**Comment by avelad** [comment]
@theodab can you help here? Thanks!

**Comment by MikeKav** [comment]
Hi, I've looked into the code differences (running uncompiled 4.8.0) and running Edge side by side with my device.

I do see this error:
[Warning] Error installing polyfill! (log.js, line 148)
ReferenceError: Can't find variable: EncryptionSchemePolyfills
install — encryption_scheme.js:26
installAll — all.js:25
(anonymous function) — ui.js:286
scanPageForShakaElements_ — ui.js:284

This is different on Edge:
log.js:148  Error installing polyfill! ReferenceError: EncryptionSchemePolyfills is not defined
    at Object.install [as callback] (encryption_scheme.js:26:5)
    at shaka.polyfill.installAll (all.js:25:18)
    at HTMLDocument.initApp (myapp.js:41:20)

I believe Edge uses native EME, whereas for my device it's probably safe to use the polyfill? (and it is set up this way in media_capabilities.js) so I'm not sure if the error matters as the playback still works on Edge

At this point in the code on my device 
<img width="1181" alt="image" src="https://github.com/user-attachments/assets/1c9e1d13-5736-436f-8c06-17dc73a0a1c8">
`const mediaKeys = await mediaKeySystemAccess.createMediaKeys();`
mediaKeys has no properties, but on Edge, this seems to be defined and have some functions attached to it:
<img width="998" alt="image" src="https://github.com/user-attachments/assets/c55f5372-2102-4ccd-884e-e14ae8a97465">

Does this look like it could be an issue? I think my next step is to see if the working version for my device (4.7.15) has similar code and if that object is correctly defined.

**Comment by MikeKav** [comment]
On device the code mediaKeySystemAccess.createMediaKeys() appears to be asserting this

// TODO(vaage): Look into the assertion below. If we do not have any drm
    // info, we create drm info so that content can play if it has drm info
    // later.
    // However it is okay if we fail to initialize? If we fail to initialize,
    // it means we won't be able to play the later-encrypted content, which is
    // not okay.

    // If the content did not originally have any drm info, then it doesn't
    // matter if we fail to initialize the drm engine, because we won't need it
    // anyway.
    return hadDrmInfo ? p : p.catch(() => {});

This is a little out of my experience area, but it seem the polyfill is not doing the right thing, I have tried using native mediaCapabilities but that does not work on my device

**Comment by MikeKav** [comment]
Part of this currently seems to be a red herring, debugging 4.7.14 I can see that 
`const mediaKeys = await mediaKeySystemAccess.createMediaKeys();`
mediaKeys has no properties as per 4.8.0

**Comment by MikeKav** [comment]
After further debugging It seems like in 4.8.0 the video element doesn't get beyond loadedmetadata on device. Not sure why this is the case. 

I spent some time reviewing: onKeyStatusesChange_ in drm_engine.js as on device the session where simple numbers compared to Edge. However the code looked to be functioning correctly in there

On device: Key status changed for session
On Edge: Key status changed for session BzBfqjxFy6Dj1SEQreP9mg==

**Comment by MikeKav** [comment]
Hi @avelad @theodab , sorry to bother you, however is there any UK offices the Shaka team work from that I could bring some hardware for debugging this issue?

**Comment by avelad** [comment]
Sorry @MikeKav, no one from the maintainers team is in the UK, sorry...

**Comment by Iragne** [comment]
Look we had to revert back our PS4 and 5 we are checking it. hope we will find the reason

**Comment by Iragne** [comment]
not in 4.8.20 but in 4.9.0

**Comment by MikeKav** [comment]
Hi @Iragne I'm not working on PlayStation myself (not since 2021), but it would be interesting to know of any areas to look that might be similar to the issue here.

I have been on holiday but we have had some time to look a little more, I have not found anything at ShakaPlayer level yet, but looking at device firmware logs we can see that the PlayReady license looks to be requested ok, and we can see what seems to be a correct license with the expected content id being stored and made ready for use in the license store.

The device uses gstreamer under the hood, there is a first-pts callback fired which normally results in an event haveEnoughData which should fire the loadedmetadata JavaScript event during loading. when using 4.7.15 this callback does occur, in 4.8.0 it doesn't. At the low level our assumption at the moment is that this means the video buffer is not being injected into the gstreamer pipeline. We don't know what would cause this difference yet

We do have a couple of hosted links that we could PM someone, however the links are geofiltered to the UK.

Unfortunately it will take a little time to have some gstreamer debugging completed, so I am going to look a little more at the ShakaPlayer differences in the preload system in 4.8.0

**Comment by Iragne** [comment]
On our side we had to set the stream.cleardecodingcache to false

**Comment by Iragne** [comment]
@MikeKav let me know if it works

**Comment by tykus160** [comment]
@Iragne `streaming.clearDecodingCache` is set to `true` by default only on PlayStation devices, so probably it won't help in the issue @MikeKav is facing.
It's been introduced because reusing of `MediaKeySystemAccess` objects is problematic on those devices and after several playbacks (i.e. due to bingewatching) they are not able to create `MediaKeys` object anymore (shaka starts to throw 6002 `FAILED_TO_CREATE_CDM` error). Clearing decoding cache after every session helped us to solve the issue on PlayStation 5.

**Comment by Iragne** [comment]
@tykus160 I was going to say the same. but we should maybe open a different bug report for PS5 and PS4.
The streaming.clearDecodingCache is set to true Clearly break the playback on our stream.
ISsue was really similar so we though it was related. We understand the patch but the reality is that it's not working on our stream AWS stream mpv2 dash. we are quite puzzle witht he next steps now happy to talk in private for that

**Comment by avelad** [comment]
@Iragne any update?

**Comment by Iragne** [comment]
@avelad  we just change the config to use the streaming.clearDecodingCache and it unblock our playback. we will share shortly information about it

**Comment by avelad** [comment]
@Iragne any update?

**Comment by agajassi** [comment]
@Iragne `+1` on your issue. That PR with new config to `clearDecodingCache` broke PlayReady playback on PlayStation for us as well. Setting it to `false` did not correct it for us unfortunately.

**Comment by MikeKav** [comment]
I've not yet tried the above, but I did notice that streaming.clearDecodingCache was implemented at a later version of player than we have been having issues with.

Stepping through today I did note that mediaKeys looks to be null. Is this ever an expected case?

**Comment by MikeKav** [comment]
I've tried testing this:

`  var config = [{
    videoCapabilities: [{
      contentType: 'video/mp4; codecs="avc1.640028"'
    }]
  }];

for (let i = 0; i < 30; i++) {
    console.log(i);
  console.log('Requesting PlayReady system access...');
    let mksa = await navigator.requestMediaKeySystemAccess(
        "com.microsoft.playready", config
    );
    console.log(mksa);
    mksa = null;
    await new Promise((resolve) => setTimeout(resolve, 2000));
}
`
This works for me on Edge, but does not work for me on the device I am using, reporting an error:
SyntaxError: Unexpected identifier 'navigator'. Expected ';' after variable declaration.

However avoiding the await functions this works
`
var config = [{
    videoCapabilities: [{
      contentType: 'video/mp4; codecs="avc1.640028"'
    }]
  }];

let resolvedCount = 0;  // Counter to track resolved promises

for (let i = 0; i < 240; i++) {
  console.log(i);
  console.log('Requesting PlayReady system access...');
  
  navigator.requestMediaKeySystemAccess('com.microsoft.playready', config)
  .then(mediaKeySystemAccess => {
    console.log(mediaKeySystemAccess);
    mediaKeySystemAccess.createMediaKeys();
    resolvedCount++;  // Increment the count for each successful promise resolution
    console.log('Number of resolved promises: ' + resolvedCount);  // Log the current count
  })
  .catch(error => {
    console.log('Argh! ' + error);
  });
}`

**Comment by joeyparrish** [comment]
`await` can only be used inside an `async function` or in some debuggers.  If you wrap the first version in an async function, it might work:

```js
async function test() {
  // ... your code here
}
test();
```

**Comment by MikeKav** [comment]
Thanks @joeyparrish 

That was the issue, this works fine on device:
```
async function test() {

  var config = [{
    videoCapabilities: [{
      contentType: 'video/mp4; codecs="avc1.640028"'
    }]
  }];

let resolvedCount = 0;  // Counter to track resolved promises

for (let i = 0; i < 240; i++) {
    console.log(i);
  console.log('Requesting PlayReady system access...');
    let mksa = await navigator.requestMediaKeySystemAccess(
        "com.microsoft.playready", config
    );
    console.log(mksa);
    mksa.createMediaKeys();
    resolvedCount++;  // Increment the count for each successful promise resolution
    console.log('Number of resolved promises: ' + resolvedCount);  // Log the current count
    mksa = null;
    await new Promise((resolve) => setTimeout(resolve, 100));
}

}
test();
```

Seems this part is ok for me

**Comment by MikeKav** [comment]
Hi @joeyparrish 

Do you have any hints on the code areas I could look at to verify that the PlayReady license request has been parsed correctly and passed into media decoding for keyid and contentkey?

**Comment by joeyparrish** [comment]
lib/media/drm_engine.js is where most DRM code lives.  Hope that helps!

**Comment by MikeKav** [comment]
I've created a sample that demonstrates the issue on my device, if anyone has time to test this on other devices to see if it works it would be appreciated. This works for me on Edge with any version number
Usage
- Unpack to web server, then use via https://yourdomain.com/index-simple.html?version=x.x.x
[simple.zip](https://github.com/user-attachments/files/17058570/simple.zip)

Working version 4.7.14, anything after 4.8.0 fails:

**Comment by MikeKav** [comment]
This simple sample on my device using 4.7.14 results in:
[Log] [Shaka Player] manifest request (myapp-simple.js, line 222)
[Log] [Shaka Player] The video has now been loaded! (myapp-simple.js, line 228)
[Log] [Shaka Player] PR license request (myapp-simple.js, line 218)
[Log] [Shaka Player] loadedmetadata received (myapp-simple.js, line 137)
[Log] [Shaka Player] loadeddata received (myapp-simple.js, line 133)
[Log] [Shaka Player] canplay received, starting playback (myapp-simple.js, line 112)
[Log] [Shaka Player] canplaythrough received (myapp-simple.js, line 117)
[Log] [Shaka Player] play received (myapp-simple.js, line 121)

This simple sample on my device using 4.8.0+ results in:
[Log] [Shaka Player] manifest request (myapp.js, line 325)
[Log] [Shaka Player] The video has now been loaded! (myapp.js, line 332)
[Log] [Shaka Player] PR license request (myapp.js, line 321)
[Log] [Shaka Player] loadedmetadata received (myapp.js, line 166)

The sample originally called play() on receiving canplay, but changing that to loadedmetadata had no effect.

**Comment by joeyparrish** [comment]
Could you please log `JSON.stringify(player.getBufferedInfo())`?

**Comment by MikeKav** [comment]
I will look at that, is there any particular point in the code or execution that would be most useful?

**Comment by MikeKav** [comment]
Edge:
'{"total":[{"start":0,"end":11.903999}],"audio":[{"start":0,"end":11.903999}],"video":[{"start":0,"end":11.999999}],"text":[]}'

Device:
"{\"total\":[{\"start\":0,\"end\":11.903999}],\"audio\":[{\"start\":0,\"end\":11.903999}],\"video\":[{\"start\":0,\"end\":12.083332}],\"text\":[]}"

**Comment by joeyparrish** [comment]
So no meaningful difference in the buffered ranges.  (The small differences we see are likely in whether they are measuring PTS or DTS.  Standards for this were unclear at one time.)

If everything is buffered from 0-11.9, and `videoElement.currentTime` is within that same range, I don't see any reason `loadeddata` would not fire.  Can you log `currentTime` under these circumstances as well?

**Comment by joeyparrish** [comment]
Properties that are helpful to know the state of:

 - mediaElement.currentTime
 - mediaElement.duration
 - mediaElement.readyState
 - mediaElement.playbackRate
 - mediaElement.paused
 - mediaElement.ended
 - player.getBufferedInfo()

https://github.com/shaka-project/shaka-player/blob/d994f7185ed301022681877491dc6b3a1d97e98b/test/test/util/waiter.js#L385-L405

These are logged by us when we get timeouts in various automated tests, to help us understand what went wrong.  The same should apply here.

**Comment by MikeKav** [comment]
I simply pasted this into the console to see:
```
   error.message += '\n' + 
       `current time: ${video.currentTime}\n` + 
       `duration: ${video.duration}\n` + 
       `ready state: ${video.readyState}\n` + 
       `playback rate: ${video.playbackRate}\n` + 
       `paused: ${video.paused}\n` + 
       `ended: ${video.ended}\n` + 
       `buffered: ${JSON.stringify(player.getBufferedInfo())}\n`;
console.log(error.message)
```
4.7.14 as an example
```
current time: 4.958311111
duration: 734
ready state: 4
playback rate: 1
paused: false
ended: false
buffered: {"total":[{"start":0,"end":15.871999}],"audio":[{"start":0,"end":15.871999}],"video":[{"start":0,"end":16.083332}],"text":[]}
```

4.8.0 was always stuck on
```
current time: 0
duration: 734
ready state: 1
playback rate: 1
paused: true
ended: false
buffered: {"total":[{"start":0,"end":11.903999}],"audio":[{"start":0,"end":11.903999}],"video":[{"start":0,"end":12.083332}],"text":[]}
```

**Comment by MikeKav** [comment]
I am currently using these streaming settings:

**Comment by avelad** [comment]
PS4 issue resolved in https://github.com/shaka-project/shaka-player/commit/58f666ba7a1bc75baa70ab612e13caf46f70b737

**Comment by avelad** [comment]
@MikeKav Can you check if applying the same solution as in https://github.com/shaka-project/shaka-player/commit/58f666ba7a1bc75baa70ab612e13caf46f70b737 solves your problem? Thanks!

**Comment by Iragne** [comment]
@avelad  we will test it too

**Comment by MikeKav** [comment]
Hi all, especially @avelad @joeyparrish @Iragne 
Firstly thanks for your help trying to address this issue. 

For historical interest, the issue seemed related to our player implementation on the device (gstreamer based). I've attached a high level summary

> When playback is started for PlayReady encrypted content of version 4.8 of ShakaPlayer the pipeline is created before the media key and license is set. This means that unlike in v4.7.15 where the pipeline is added and then the drm is set which allows the object to be set on the underlying gstreamer pipeline, no such pipeline exists when the drm object is added. Subsequently when the pipeline is created/attached there was no available Key IDs (kids) and so the drm object is not set to the lower levels. Thus when a buffer is sent down the pipeline to be decrypted no DRM object is available/configured.
> 
> 
> From looking at Shaka logs we didn't see an obvious reason why there is a different ordering of these actions between versions, although it appears that the async calls may not be being actioned quite in the same order compared to earlier versions.

We now have successful playback on PlayReady content from 4.8.0 to 4.11.11, so happy days. During testing we have seen an case where Shaka is taking ~12 seconds to transition from loadedmeta to loadeddata for certain content, but that is a separate issue we will look into
